### PR TITLE
SWD Tandem Domain Alignment: Slice Token Distribution Matching

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The intermediate slice token distributions of tandem samples vs single-foil samples are misaligned — the model routes them through different feature pathways. This distributional mismatch hurts OOD NACA6416 generalization: the model has learned that "tandem-looking slice distributions" go one way, and NACA6416 in tandem looks different from all seen tandem configs.

**Proposed:** Add an auxiliary loss minimizing the Sliced Wasserstein Distance (SWD) between slice token distributions of tandem and single-foil samples. This forces domain-agnostic intermediate representations at the slice level.

SWD is computationally cheap: O(K × N log N) where K=50 random projections, N = number of slice tokens. Negligible overhead. No adversarial training needed — direct optimal transport distance.

Reference: Lee et al. "Sliced Wasserstein Discrepancy for Unsupervised Domain Adaptation" (CVPR 2019, arXiv:1904.11430).

## Instructions

**Step 1 — Add SWD function:**
```python
def sliced_wasserstein_distance(x, y, n_projections=50):
    """
    x: [N_x, d] — slice tokens from tandem samples
    y: [N_y, d] — slice tokens from single-foil samples
    Returns scalar SWD loss.
    """
    d = x.shape[-1]
    directions = F.normalize(torch.randn(n_projections, d, device=x.device), dim=-1)
    x_proj = x @ directions.T   # [N_x, n_proj]
    y_proj = y @ directions.T   # [N_y, n_proj]
    x_sorted = x_proj.sort(dim=0).values
    y_sorted = y_proj.sort(dim=0).values
    # Resample to same length
    n = min(x_sorted.shape[0], y_sorted.shape[0])
    x_sorted = x_sorted[:n]
    y_sorted = y_sorted[:n]
    return ((x_sorted - y_sorted) ** 2).mean()
```

**Step 2 — Add CLI flags:**
```python
parser.add_argument('--swd_align', action='store_true', help='Enable SWD domain alignment')
parser.add_argument('--swd_weight', type=float, default=0.01, help='SWD loss weight')
parser.add_argument('--swd_n_proj', type=int, default=50, help='Number of random projections')
parser.add_argument('--swd_start_epoch', type=int, default=20,
                    help='Delay SWD loss until primary features established')
```

**Step 3 — Apply in training loop.** After the final TransolverBlock, extract slice tokens and compute SWD:
```python
if args.swd_align and epoch >= args.swd_start_epoch:
    # Need access to slice tokens from the last TransolverBlock
    # These are the intermediate representations before the decoder
    tandem_mask = is_tandem.bool()  # identify tandem samples in batch
    if tandem_mask.any() and (~tandem_mask).any():
        tandem_tokens = slice_tokens[tandem_mask].reshape(-1, n_hidden)
        single_tokens = slice_tokens[~tandem_mask].reshape(-1, n_hidden)
        swd = sliced_wasserstein_distance(tandem_tokens, single_tokens, args.swd_n_proj)
        total_loss = total_loss + args.swd_weight * swd
```

**Important implementation notes:**
- You need to extract `slice_tokens` from the model's forward pass. Look at how `TransolverBlock.forward()` returns slice tokens (the attention-weighted features). You may need to add a hook or modify the forward to return them.
- The `is_tandem` flag should already exist in the training loop (used by tandem_ramp). Use the same mechanism.
- If both `tandem_mask.any()` and `(~tandem_mask).any()` aren't both True in a batch, skip SWD for that batch (log how often this happens).
- Log `swd_loss` value to W&B every 20 steps to monitor convergence.

**Step 4 — Run 2 configs, 2 seeds each** using `--wandb_group askeladd-swd-align`:

**Config A — weight=0.01, start_epoch=20:**
```bash
cd cfd_tandemfoil && python train.py --agent askeladd \
  --wandb_name "askeladd/swd-w0.01-e20" --wandb_group askeladd-swd-align \
  --swd_align --swd_weight 0.01 --swd_n_proj 50 --swd_start_epoch 20 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias \
  --seed 42
```
(repeat with `--seed 73`)

**Config B — stronger alignment, weight=0.05:**
Same command with `--swd_weight 0.05` and `--wandb_name "askeladd/swd-w0.05-e20"`. (Repeat with `--seed 73`.)

Run all 4 in parallel. Report p_in, p_oodc, p_tan, p_re per config (2-seed avg).

## Baseline

| Metric | Baseline (2-seed avg) |
|--------|----------------------|
| p_in | 13.05 |
| p_oodc | 7.70 |
| **p_tan** | **28.60** |
| p_re | 6.55 |

Baseline W&B runs: d7l91p0x (seed 42, p_tan=28.9), j9btfx09 (seed 73, p_tan=28.3)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Summary

SWD domain alignment at the slice token level does **not** improve tandem transfer. The low-weight config (w=0.01) is essentially neutral; the higher weight (w=0.05) actively degrades all metrics.

### Individual Run Metrics (Surface MAE p, denormalized)

| Config | Seed | W&B Run | p_in | p_tan | p_oodc | p_re | Best Epoch |
|--------|------|---------|------|-------|--------|------|------------|
| w=0.01 | 42 | `mrqnyv82` | 12.8 | 28.7 | 7.7 | 6.5 | 150 |
| w=0.01 | 73 | `uwphs04i` | 13.4 | 28.8 | 7.8 | 6.6 | 151 |
| w=0.05 | 42 | `ukxu3q8i` | 13.4 | 29.8 | 7.8 | 6.5 | 151 |
| w=0.05 | 73 | `qa84ir9x` | 13.4 | 29.5 | 8.2 | 6.7 | 150 |

### 2-Seed Averages vs Baseline

| Config | p_in | p_tan | p_oodc | p_re |
|--------|------|-------|--------|------|
| **Baseline** | **13.05** | **28.60** | **7.70** | **6.55** |
| SWD w=0.01 | 13.10 (+0.4%) | 28.75 (+0.5%) | 7.75 (+0.6%) | 6.55 (0%) |
| SWD w=0.05 | 13.40 (+2.7%) | 29.65 (+3.7%) | 8.00 (+3.9%) | 6.60 (+0.8%) |

### Train Commands

```bash
# Config A: swd_weight=0.01
CUDA_VISIBLE_DEVICES=0 python train.py --agent askeladd --wandb_name "askeladd/swd-w0.01-e20" --wandb_group askeladd-swd-align \
  --swd_align --swd_weight 0.01 --swd_n_proj 50 --swd_start_epoch 20 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias --seed 42

# Config B: swd_weight=0.05 (same as above with --swd_weight 0.05)
```

### Peak Memory

~43 GB VRAM per run (same as baseline — SWD overhead negligible)

### W&B Run IDs

- `mrqnyv82` (w=0.01, seed 42), `uwphs04i` (w=0.01, seed 73)
- `ukxu3q8i` (w=0.05, seed 42), `qa84ir9x` (w=0.05, seed 73)
- W&B group: `askeladd-swd-align`

### What happened

**The hypothesis did not work.** Aligning slice token distributions between tandem and single-foil samples via SWD does not improve tandem transfer performance.

Key observations:
1. **w=0.01 is neutral** — the loss is too small to meaningfully shift representations. p_tan essentially matches baseline (28.75 vs 28.60, within seed variance).
2. **w=0.05 actively hurts** — forcing alignment at this strength degrades all metrics, especially p_tan (+3.7%) and p_oodc (+3.9%). This suggests the distributional difference between tandem and single-foil slice tokens is *informative*, not harmful — the model legitimately needs different routing for multi-foil configurations.
3. **SWD loss converged** (final value ~0.115) — the mechanism worked correctly, the alignment was being enforced. The problem is that the alignment itself is counterproductive.
4. The tandem-specific routing through slice tokens appears to encode physically meaningful multi-body interaction effects (wake interference, gap flow dynamics) that are lost when forced to match the single-foil distribution.

### Note on verification NaN bug

All 4 runs show `Pipeline MAE surf_p: nan` in the verification section at the end. This is a pre-existing issue in the verification code (the `visualize()` function also fails with missing positional arguments) — not related to SWD changes. The main training metrics reported by the model during training are valid.

### Suggested follow-ups

1. **Selective alignment**: Instead of aligning all slice tokens, align only a subset (e.g., the tokens most similar between domains, or only specific attention heads). Some heads may encode domain-invariant physics while others encode domain-specific geometry.
2. **Contrastive approach**: Rather than minimizing SWD, use a contrastive loss that pulls together "same physics, different domain" tokens while pushing apart "different physics" tokens. This preserves domain-specific information while encouraging shared physics representations.
3. **Gradient reversal**: Instead of SWD, try a domain discriminator with gradient reversal layer (DANN-style). This can learn which features to align vs. preserve, rather than naively aligning everything.
4. **Shared decoder constraint**: Instead of aligning intermediate representations, enforce that the decoder weights used for tandem predictions are similar to those used for single-foil — targeting the output mapping rather than intermediate features.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2175 because Results are clear: SWD domain alignment does not help tandem transfer. w=0.01 is neutral (p_tan 28.75 vs baseline 28.60) and w=0.05 actively degrades all metrics (+3.7% p_tan). The student's analysis nails it — the distributional difference between tandem and single-foil slice tokens encodes physically meaningful multi-body interaction effects (gap flow, wake interference) that the model legitimately needs. Forcing alignment destroys that signal. Good clean experiment, well-executed. Closing as a clear dead end..
